core/src/upgrade.rs

```diff
@@ -134,6 +135,26 @@ impl<T: AsRef<[u8]>> ProtocolName for T {
     }
 }
 
+pub struct DisplayProtocolName<N>(N);
+impl<N: ProtocolName> Display for DisplayProtocolName<N> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        for byte in self.0.protocol_name() {
+            if (b' '..=b'~').contains(byte) {
+                f.write_char(char::from(*byte))?;
+            } else {
+                write!(f, "<{:02X}>", byte)?;
+            }
+        }
+        Ok(())
+    }
+}
+
+#[test]
+fn protocol_name_display() {
+    assert_eq!(DisplayProtocolName("hello").to_string(), "hello");
+    assert_eq!(DisplayProtocolName("hellö/").to_string(), "hell<C3><B6>/");
+}
```

Comment:
Can we put this with the other tests at the bottom please in a `mod tests`?

Reply:
ah, I looked at some `tests.rs` but found it irrelevant, so placed the test next to the thing; recently I began doubting the dogma of “`mod tests` at the bottom” for such things because they serve to clearly document the intended behaviour right where people read the code

Reply:
I don't mind much where the tests are but I want them to be feature-gated on `cfg(test)` so we don't compile them into binaries of our users.

Reply:
ah, good point: we have a deal :-)

Reply:
Wow, that was interesting: I checked what was needed, and indeed `#[test]` alone already hides the function in normal builds. This is the expansion:
```rust
    extern crate test;
    #[cfg(test)]
    #[rustc_test_marker = "upgrade::protocol_name_display"]
    pub const protocol_name_display: test::TestDescAndFn = test::TestDescAndFn {
        desc: test::TestDesc {
            name: test::StaticTestName("upgrade::protocol_name_display"),
            ignore: false,
            ignore_message: ::core::option::Option::None,
            compile_fail: false,
            no_run: false,
            should_panic: test::ShouldPanic::No,
            test_type: test::TestType::UnitTest,
        },
        testfn: test::StaticTestFn(|| test::assert_test_result(protocol_name_display())),
    };
    fn protocol_name_display() {
        ...
    }
```
So thanks to our exchange I learnt that `#[cfg(test)]` is only needed for hiding test-related support functions and infrastructure.

Reply:
Are you sure `extern crate test` is not being dragged in?

Reply:
Hmm, after lunch I followed this rabbit hole for a little while longer and compared the liblibp2p_core.rlib for three cases:

1. without the test (commented out)
2. test with only the `#[test]` attribute (120 bytes size increase, caused by a 16 byte increase and other changes in `lib.rmeta`)
3. with additional `#[cfg(test)]` attribute (106 bytes size increase, caused by a 104 byte increase in `lib.rmeta`)

As the metadata format is somewhat obscure, I decided to stop here. To my knowledge, `extern crate test` only brings a declared dependency explicitly into scope, which has been unnecessary since the 2015 edition (IIRC). Judging by the absence of the string `protocol_name_display` in the rlib, I’d leave the code as is, unless you (or others) have a strong opinion.

Reply:
I am leaning towards add a `mod tests` block:

1. Based on your investigation (thanks for doing that!), it seems to reduce code size.
2. It is consistent across the codebase.

I don't have a strong opinion on where the `mod tests` block sits although we tend to put it at the bottom of the file.

The argument that it should be next to the function is definitely valid, however I think the more common case is that you want to read the production code, not the test code.

If you want to actually co-locate it, you could make it a rustdoc test :)

Reply:
That is a brilliant idea, in particular, because it found a usability bug (no public constructor).